Documentation/botr/readytorun-overview.md

```diff
@@ -0,0 +1,335 @@
+
+
+## Object Creation - NEWOBJ / NEWARR
+
+Object allocation is always done by a helper call that allocates the uninitialized object memory (but does initialize the type information `MethodTable` pointer), followed by calling the class constructor. There are a number of different helpers depending on the characteristics of the type (does it have a finalizer, is it smaller than a certain size, ...).
+
+We will defer the choice of the helper to use to allocate the object to runtime. For example, to create an instance of `SubClass` the code would be:
+
+        CALL [NEWOBJ_SUBCLASS]
+    	MOV RCX, RAX  // EAX holds the new object		
+		// If the constructor had parameters, set them
+	    CALL SUBCLASS_CONSTRUCTOR
+
+    .data // In the data section
+    NEWOBJ_SUBCLASS: RUNTIME_ENTRY_FIXUP // One per type
+
+where the `NEWOBJ_SUBCLASS` would be fixed up using the standard lazy technique.
+
+The same technique works for creating new arrays (NEWARR instruction). 
+
+
+## Type Casting - ISINST / CASTCLASS
+
+The proposal is to use the same technique as for object creation. Note that type casting could easily be a case where VSD techniques would be helpful (as any particular call might be monomorphic, and thus caching the result of the last type cast would be a performance win. However this optimization is not necessary for version resilience. 
+
+
+## GC Information for Types	
+
+To do its job the garbage collector must be able to take an arbitrary object in the GC heap and find all the GC references in that object. It is also necessary for the GC to 'scan' the GC from start to end, which means it needs to know the size of every object. Fast access to two pieces of information is what is needed. 
+From a versioning perspective, the fundamental problem with GC information is that (like field offsets) it incorporates information from the entire inheritance hierarchy in general case. This means that the information is not version resilient.
+
+While it is possible to make the GC information resilient and have the GC use this resilient data, GC happen frequently and type loading happens infrequently, so arguably you should trade type loading speed for GC speed if given the choice. Moreover the size of the GC information is typically quite small (e.g. 12-32 bytes) and will only occur for those types that cross versioning bubbles. Thus forming the GC information on the fly (from a version resilient form) is a reasonable starting point. 
+
+Another important observation is that the `MethodTable` contain other very frequently accessed data, like flags whether the `MethodTable` represents an array, or pointer to parent type. This data tend to change a lot with evolution of the runtime. Thus, generating `MethodTables` at runtime will solve a number of other versioning issues in addition to the GC information versioning.
```

Comment:
Another important observation is that `MethodTable` contains other very frequently accessed data, like flags indicating whether the `MethodTable` represents an array, or pointer to parent type. This data tends to change a lot with the evolution of the runtime. Thus, generating method tables at runtime will solve a number of other versioning issues in addition to the GC information versioning.